### ISSUE TITLE
Проверка валидности строки как команды.

### ISSUE THREAD

**Issue by ivanbalashov239** [uzyovoys/aggregate#33]
предположим есть модуль отложенных задач, таймеров, `$Text через $Number минут`  должен запустить $Text в нужный момент. Правильно ли я понимаю, что его нужно передать как result здесь:

``` java
vertx.eventBus().publish("asr.result", result);
```

И главное, как проверить валидность $Text как команды агрегата?

**Comment by morfeusys** [comment]
Не совсем понятно... asr.result - это событие, которое посылает какой-либо модуль распознавания речи.
Если у вас есть модуль, который понимает команды вида `$Text через $Number минут*`, то ваш модуль должен реализовывать задачу выполнения отложенных задач. Если вам нужны только какие-то валидные задачи - то либо модуль должен проверять $Text, либо нужно записать шаблон типа `$MyCommand`, который содержит альтернативы всех валидных задач.

**Comment by ivanbalashov239** [comment]
@Morfeusys вижу, что сумбурно написал, идея в следующем:
1) у меня есть 2 модуля: первый умеет выполнять команду, второй умеет описанное в предыдущем вопросе, то есть ставит таймер.
2)вопрос в том, как второй модуль может спросить агрегат, считает ли агрегат некоторую строку понятной/валидной командой и в случае положительного ответа отдать её на выполнение(задержку выставить не проблема).
3)Я решил, что если послать asr.result с некоторой строкой из любого модуля, то агрегат обработает её как входящую команду, отдаст какому то функциональному модулю, согласно своим правилам.
4) пример, есть модуль который понимает команду "усни", модуль таймер услышав команду "усни через 5 минут"  должен понять, что команду "усни" нужно отдать агрегату на выполнение через 5 минут. соответственно модуль таймера должен узнать есть ли такая команда у агрегата в нынешней конфигурации, и в нужный момент запустить её.

**Comment by morfeusys** [comment]
Я вас понял чуть позже, чем написал свой коммент :)

Дело в том, что по текущей парадигме модуль - это только слушатель события. Связи "в обратную сторону" с агрегатом у него нет. Это связано с тем, что хочется максимально развязать компоненты, что дает возможности по реализации модулей на самых разных языках программирования. Я считаю что эта "мультиязычность" даст проекту многое.

С точки зрения vertx конечно есть возможность создавать синхронные сервисы, к которым можно обратиться из любого кода. Но это накладывает некоторые ограничения и усложняет API.

Но это не основная проблема.  Проблема еще заключается в том, как сейчас идет выбор лучшего соответствия (матчинга) текста некоторому шаблону. Сейчас это быстрый алгоритм, который учитывает лишь количество символов, точно совпавших с той частью шаблона, которая описана скажем так "четко". Поэтому если скажем есть модуль который имеет шаблон `* мама мыла раму *`, то ваш модуль с шаблоном `$Text через $Number минут*` получит управление только если ваш $Number будет состоять более чем из 2-х символов.

**Comment by ivanbalashov239** [comment]
@Morfeusys $Number будет результатом распознавания, поэтому будет словом-числительным, больше 2 символов наверняка, А в дусе такой же матчинг? Просто по идее агрегату доступны значительно большие ресурсы, а требования к энергосбережению меньше, можно не так ужиматься.
мультиязычность это хорошо, конечно, жаль что это мешает получать ответ от агрегата. Но послать asr.result из произвольного модуля это же не мешает? агрегат воспримет это как результат  распознавания и будет его обрабатывать как обычно, Правильно?

**Comment by morfeusys** [comment]
@seniorivn нет, это не факт, Все зависит от системы распознавания речи. Гугл может вернуть число и в виде цифр.

Ужиматься всегда нужно, ибо Агрегат работает не только на мощных ПК, в этом его идеология. Он может работать и на слобосильных Raspberry pi - скоро покажу ))

**Comment by ivanbalashov239** [comment]
@Morfeusys ладно, с малиной понятно надо ужиматься, хотя тут лучше сказать оптимизировать, но это всё больше лирика а главное, как такой функционал реализовать то? Это остро необходимо, ведь это не единственный сценарий, совершенно нормально, если один функциональный модуль использует возможности другого.

**Comment by morfeusys** [comment]
@seniorivn это можно назвать нормальным, если мы точно знаем, что одному модулю требуется другой. А если такая зависимость присутствует на архитектурном уровне модуля, то тогда никакие хитрые решения этого вопроса не нужны - просто пишутся правильные шаблоны и все.

Конечно вопрос мультиязычного, слабосвязанного и реактивного фреймворка голосового ассистента с возможностью декларативной связи между модулями - это прекрасная вещь, но она требует плотной работы в этом направлении.

**Comment by morfeusys** [comment]
@seniorivn Возвращаясь к этому вопросу, могу сказать что меня посетила хоть и запоздалая, но светлая мысль! А почему бы просто не послать событие-сообщение агрегату из любого модуля, кому нужно сматчить на глобальном контексте строку, и получить в ответ асинхронно результат? Это решает вопрос с разными языками программирования, так как система событий от них независима. 

Не знаю как это сразу в голову не пришло )) Планирую на версию 0.1.0 агрегата.

**Comment by ivanbalashov239** [comment]
@Morfeusys ну вот и отлично, я и не понял где ты проблему увидел, надо было вдуматься, раньше подсказал бы.
